Round depth values to 16-bit precision to match the PSP's 16-bit depth buffer

hrydgard opened the pull request:
As most cases where this will matter (like in Phantasy Star #3777) are simply screenspace 2D polygons, we do the rounding in the vertex shader. We can avoid tricks like #7912 this way. Also, modifying Z in the fragment shader is very expensive on some GPUs so we'd like to avoid that.

Needs testing on all applicable platforms. Not sure if the performance hit for this extra work will be noticeable.

hrydgard committed:
Failed quick attempt at Z rounding in the vshader

hrydgard committed:
Round Z in the vertex shader to match a 16-bit Z buffer.

However this inexplicably does still not solve the Phantasy Star issue!

hrydgard committed:
Correct scale factors for through mode depth rounding.

LunaMoo commented:
With that third commit it fixed PSP1 and Heroes Phantasia:), hopefully also some other games on depth issues list:).

hrydgard committed:
Fix typo. This breaks it again though, strangely.

hrydgard committed:
Implement vertex shader depth rounding in DX9

hrydgard commented:
It did, but it also introduced a typo which accidentally nudged it enough to be able to work, apparently :( Now that I've fixed that, it's broken again. Although does seem to work in DX9..

unknownbrackets commented on GPU/GLES/GLES_GPU.cpp:
```diff
@@ -206,8 +206,8 @@ static const CommandTableEntry commandTable[] = {
-	{GE_CMD_VIEWPORTZ1, FLAG_FLUSHBEFOREONCHANGE | FLAG_EXECUTEONCHANGE, 0, &GLES_GPU::Execute_ViewportType},
-	{GE_CMD_VIEWPORTZ2, FLAG_FLUSHBEFOREONCHANGE | FLAG_EXECUTEONCHANGE, 0, &GLES_GPU::Execute_ViewportType},
+	{GE_CMD_VIEWPORTZ1, FLAG_FLUSHBEFOREONCHANGE | FLAG_EXECUTEONCHANGE, DIRTY_DEPTHRANGE, &GLES_GPU::Execute_ViewportType},
+	{GE_CMD_VIEWPORTZ2, FLAG_FLUSHBEFOREONCHANGE | FLAG_EXECUTEONCHANGE, DIRTY_DEPTHRANGE, &GLES_GPU::Execute_ViewportType},
```
I could be wrong, but doesn't Execute_ViewportType still need to dirty?

-[Unknown]

unknownbrackets commented:
Technically it rounds out 0.625 iirc... not that it matters that much.

-[Unknown]

unknownbrackets commented:
Phantasy Star Portable 1 works for me actually?

-[Unknown]

hrydgard committed:
Properly dirty the uniform in Execute_ViewportType

hrydgard commented:
"rounds out"? should we add a small offset?

Oh, it does? The text does not show up for me in the latest commit.. GTX780 or something.

daniel229 commented:
So far it fixes missing text in Photokano and Taiko no Tatsujin DX.Phantasy Star Portable 2 JP is still broken.

LunaMoo commented:
Heroes Phantasia get's broken again with typo corrected althrough maybe slightly less in some places.

unknownbrackets commented on GPU/GLES/GLES_GPU.cpp:
```diff
@@ -1012,6 +1012,12 @@ void GLES_GPU::Execute_FramebufType(u32 op, u32 diff) {
+	case GE_CMD_VIEWPORTZ1:
+	case GE_CMD_VIEWPORTZ2:
+		shaderManager_->DirtyUniform(DIRTY_DEPTHRANGE);
+		break;
```
I'm nitpicking now, but I would vote Execute_FramebufZType or something, though probably viewport commands are cheap enough it doesn't matter...

-[Unknown]

hrydgard replied:
Done

unknownbrackets commented:
Sorry, for posterity I meant "rounds at" - that is, 0.625 rounds up, 0.624 rounds down.  It's like this with all the coordinate values (x, y, and z.)

It turns out, on my 9800 GT, I'm getting a slightly lower value than it should be...

-[Unknown]

daniel229 commented:
Puyo Puyo Fever 2 is fixed too.

hrydgard committed:
Subtracting an offset of 0.5 works on my nVidia and I have no idea why, this makes no sense.

hrydgard commented:
@daniel229  thanks for your testing! I added another small tweak that fixes Phantasy Star 1 for me but I don't understand why. It's super sensitive..

hrydgard committed:
experiment: Add a different offset that unknown found during testing, plus remove rounding of through mode depth (which should already be rounded)

hrydgard committed:
Get rid of a division in the vertex shader by precomputing the inverse.

hrydgard committed:
Oops, didn't finish the D3D code

hrydgard commented:
Alright, done for now I think - it seems to work now. @daniel229 , can you test your games again?

LunaMoo commented:
Any clue why would adding

    WRITE(p, "  z /= u_depthRange.z;\n");

fix Heroes Phantasia?:3 In GE debugger under depth,the second value(after /) get's over 2 times higher than it normally is, but ends up consistent for all character parts while the flicker shows up when there are different numbers for different parts(althrough the difference is very small like 13541.000977 and 13541.004883).

hrydgard commented:
That probably reduces precision enough that it all ends up being the same. That said, now that we do this rounding, two depth values that close shouldn't really still happen, although apparently it's still possible somehow, possibly when the viewport Z settings differ...

hrydgard committed:
Minor optimization

LunaMoo commented:
Viewport 1&2 under settings seems same for all parts, so maybe it's something else happening.
 Interesting I made myself a hackish build for that game with z /= 4000.0; and it also seems to fix another ~ #6531 maybe some games just need lower precision?(Original typo breaks that game through:p.)

 Anyway unless I can help, I'm not gonna bug anyone about games, so thanks for your work:).

sum2012 commented:
Well done

hrydgard commented:
@sum2012  thanks. If you feel like doing some testing to make sure this is all right, please test through your most graphically troublesome games with this :)  I will wait a day or so for test results before I merge.

ikoul commented:
Did a basic test (OpenGL renderer and whatever existing settings I had) of some games that were mentioned in relation to #7912. The shadow issues mentioned in #5711 and #5769 didn't appear to come back. Shadows in FF:Type 0 mentioned in the comments on #7912 looked fine too. This change set is looking pretty good so far.

daniel229 commented:
With the latest change, Puyo Puyo Fever 2,Taiko no Tatsujin DX and Phantasy Star Portable 2 JP still work,but Photokano not work.

daniel229 commented:
https://github.com/hrydgard/ppsspp/commit/5dbc8abd94b3727858776985cc7aa9b95b7ab3e0 broke photokano.

daniel229 commented:
Also fixes Power Stone Collection and NBA 2013.

unknownbrackets commented:
@daniel229 does changing `+ 0.375` to something else, like `+ 0.350` or `+ 0` work for Photo Kano?  Just trying to understand the problem there.  It might be some subtle rounding issue...

-[Unknown]

daniel229 commented:
+ 0 works for Photo Kano

hrydgard commented:
If +0 works for all games, let's just use that. It seems to be very hard to completely match the PSP's transform pipeline so we'll have to be happy enough with that, for now..

daniel229 commented:
All these games work before and still work in using <code>+ 0</code>
Phantasy Star Portable 2 JP
Taiko no Tatsujin DX 
Power Stone Collection 
NBA 2013
Photo Kano
Puyo Puyo Fever 2

hrydgard commented:
Thanks Daniel. I will merge tonight, with +0.

daniel229 commented:
By the way,this branch may breaks other games like Suto Mani.

hrydgard commented:
@daniel229 Oh, it does? Please report anything it breaks. It shouldn't.

daniel229 commented:
Just black screen,only video and OSK show up in Suto Mani.

daniel229 commented:
Gran Turismo is also black screen

daniel229 commented:
Broken in OpenGL,not in D3D9.

hrydgard commented:
Thanks, will have to look into that before we can merge then. This branch should not break things.

daniel229 commented:
D3D9 fixes Hokuto no Ken Raoh Gaiden Ten no Haoh,not fixes Phantasy Star Portable 2 JP,breaks Three Kingdom 9.

hrydgard committed:
Fix GT and hopefully other games that got a black screen

hrydgard committed:
Also apply the GT fix in D3D, doh

hrydgard commented:
OK, that takes care of Gran Turismo and hopefully Suto Mani as well, can you test @daniel229 ?

hrydgard committed:
Change the offset to 0.0

daniel229 commented:
Great, Suto Mani is good now.

daniel229 commented:
Three Kingdom 9 is still blackscreen in D3D9.

unknownbrackets commented on GPU/Directx9/VertexShaderGeneratorDX9.cpp:
```diff
@@ -279,6 +284,18 @@ void GenerateVertexShaderDX9(int prim, char *buffer, bool useHWTransform) {
 
+	if (!gstate.isModeThrough()) {
+		// Apply the projection and viewport to get the Z buffer value, floor to integer, undo the viewport and projection.
+		// Not completely sure this is 100% right under DX9 as the Z range is different...
```
I think we have to change the scale here?  For example, if scale and center were 32767.5 (perfectly normal values, btw) then in OpenGL:

z [-1,1] * 32767.5 + 32767.5
=> [0, 65535]

But in Direct3D (this is after the projection multiply):

z [0,1] * 32767.5 + 32767.5
=> [32767.5, 65535]

This will lose an extra bit worth of integer precision.

So what we want is really:

z * (scale * 2) - (65535.0 / 2 + center)

I think?

-[Unknown]

hrydgard replied:
Yes, that seems right, I was just too tired to figure it out. Will apply. Also it looks like we can easily do that by pre-adjusting the u_depthRange constant values.

hrydgard committed:
Adjust D3D depth calculations to get them to match OpenGL as close as I can get it.

unknownbrackets commented on GPU/Directx9/ShaderManagerDX9.cpp:
```diff
@@ -475,6 +479,26 @@ void ShaderManagerDX9::VSUpdateUniforms(int dirtyUniforms) {
+		// Given the way we do the rounding, the offset is probably mostly irrelevant as we cancel
+		// it afterwards anyway.
+		viewZScale *= 2.0f;
+		viewZCenter -= 32768.0f;
```
Hmm, isn't the offset definitely relevant for rounding?  It's basically where it rounds, and it affects the value results, no?

For example, consider:

Input: 0.1
Scale: 1000
Center: 99.5

0.1 * 1000 + 99.5 = 199.5
Floor: 199
(199 - 99.5) / 1000 = 0.0995

Now offset: 100

0.1 * 1000 + 100 = 200
Floor: 200
(200 - 100) / 1000 = 0.1

Definitely different, no?

The transformation we apply on this value is (z * 0.5) + 0.5.  This coverts the range [-1, 1] to [0, 1].

32768 is half of 65536, but that's not relevant here is it?  The range of this value is [0, 65535].

-[Unknown]

hrydgard replied:
I mistyped, I mean that the *integer* part of the offset is irrelevant.

Using 0.5 or 32767.5 here as the offset causes the depth in Phantasy Star to come out as 38023.496094 , that is, 0.5 off from the value we want. So for some reason that's not the right thing.

hrydgard committed:
This makes slightly more sense than the last commit.

hrydgard commented:
Okay, anyone feel like doing some final testing or should we go ahead and merge and just see what the fallout is? :-)

daniel229 commented:
Seems good,the last commit fixes Three Kingdom 9 and breaks Hokuto no Ken Raoh Gaiden Ten no Haoh agains in D3D9.

Depth range hack is broken in both OpenGL and D3D9.

hrydgard commented:
Argh yes, the depth range hack is not really compatible with this. However, I would expect it to not break entirely, but just work a bit worse...

dan83sama commented:
In software mode menus show prefect, and Z = 1,

http://i.imgur.com/j6TgCG5.png
